### PR TITLE
fix(llm-drivers): preserve streamed tool calls and 429 recovery

### PR DESCRIPTION
**Pull request by houko**
Three driver-layer defects, two of which silently corrupt a turn rather than failing it.

## Streamed tool calls were truncated by a delta for an earlier index

The OpenAI streaming accumulator grew its slot vector with a bare `tool_accum.resize_with(idx + 1, …)`.
`resize_with` shrinks as readily as it grows, so a delta carrying a lower index than one already seen truncated the vector and permanently discarded every later tool call — after `ToolUseStart` and `ToolInputDelta` for those calls had already been emitted to the consumer.

If that index then received a further delta, the slot was recreated as a default: `start_emitted` was false again, producing a duplicate `ToolUseStart`, and the accumulated argument prefix was gone, leaving JSON that fails to parse and degrades to `malformed_tool_input`.

Providers are not required to order `index` monotonically, and the driver's own test suite already covers a late field arriving for an earlier index.
Growth is now one-directional (`if tool_accum.len() < idx + 1`), keeping the single-allocation sparse growth that introduced the regression.

## A recovered 429 left a cross-process lockout nothing ever cleared

The 429 handler wrote a shared lockout file before the driver's own retry, so a rate limit the driver successfully recovered from still left a five-minute lockout on disk.
The file is shared across processes, so the next driver instance — a daemon fork, `librefang agent chat`, a cron fire — failed `pre_request_check` with `RateLimited` before opening a socket, for a limit that no longer applied.

A new `note_429_from_headers` records the metric and parses `Retry-After` without writing a lockout, and the recovery path uses it.
`record_429_from_headers` delegates to the same helper, so metric semantics are unchanged for the paths that genuinely need to persist a lockout.

## claude-code driver dropped the Anthropic cache token buckets

`ClaudeUsage` did not deserialize `cache_creation_input_tokens` or `cache_read_input_tokens`, so `input_tokens` reported new input only.
That violates the documented `TokenUsage` contract and understates billed input on every cache-hit turn, which propagates into metering and budget accounting.

Both buckets are now deserialized with `#[serde(default)]` and folded into `input_tokens` while being carried through as subsets — byte-for-byte the normalisation `anthropic.rs` already applies, so the two providers agree.

## Notes

Comment prose in the touched 429 blocks of `anthropic.rs` and `openai.rs` was unwrapped from hard column wraps to one sentence per line, per the repo's prose rule for paragraphs already being edited.
`anthropic.rs`, `gemini.rs` and `shared_rate_guard.rs` are in the diff because they are callers of the changed rate-guard helper, not as independent changes.

## Verification

`cargo clippy --workspace --all-targets -- -D warnings` clean, and `cargo fmt --all --check` and `scripts/check-changelog-attribution.py` clean.

One caveat worth recording, because it produced a false result first time round: these worktrees initially shared one `CARGO_TARGET_DIR`, and cargo served one worktree another's `librefang-*` artifacts — which manufactured an error in a file this branch never touched, and could equally have manufactured a pass.
Every clippy result quoted here was re-run after deleting the workspace-member artifacts, so all 29 workspace crates were genuinely recompiled from this tree.

Full test suites run in CI.

### CODE REVIEW

File: crates/librefang-llm-drivers/src/drivers/openai.rs [hunk 1]
```diff
@@ -2514,7 +2522,10 @@ impl LlmDriver for OpenAIDriver {
 
                                 // Grow to a sparse provider-supplied index in
                                 // one allocation rather than one push per slot.
-                                tool_accum.resize_with(idx + 1, StreamedToolCall::default);
+                                // Growth has to stay one-directional: `resize_with` also shrinks, so a delta for a lower index — a late `id`, a trailing `arguments` fragment, or two entries in one frame in descending order — would truncate the slots already accumulated above it and silently drop a parallel tool call the model asked for.
+                                if tool_accum.len() < idx + 1 {
```
Comment (houko) [inline crates/librefang-llm-drivers/src/drivers/openai.rs:2526]: Keeping the earlier slot alive also makes its late `arguments` fragment *emittable*, and `ToolInputDelta` carries no tool id — so consumers attribute it to the wrong call.

Before this change a trailing fragment for index 0 landed on a slot that `resize_with` had just truncated and recreated as a default, so `start_emitted` was `false` and the emission block below was skipped. Now slot 0 survives with `start_emitted == true`, so line 2573 fires a `ToolInputDelta { text }` for index 0 *after* index 1 has already started.

`crates/librefang-api/src/openai_compat.rs:537` stamps every `ToolInputDelta` with `index: state.tool_index - 1`, i.e. the most recently started call:

```
idx0 start (+ {"x":1)   -> chunk index 0
idx1 start (+ {"y":2})  -> chunk index 1
idx0 trailing }         -> chunk index 1   <-- wrong
```

An OpenAI-compat client then reconstructs index 1's arguments as `{"y":2}}` (invalid JSON) and index 0's as `{"x":1` (truncated), even though this driver's own `response.tool_calls` is now correct. `crates/librefang-cli/src/tui/mod.rs:1655` has the same single-buffer attribution.

The final `ToolUseEnd` already carries the complete `input`, so the cheapest fix is to emit the mid-stream delta only while `idx` is the highest-started slot and let earlier-index fragments be delivered by `ToolUseEnd`; the alternative is adding the tool id to `StreamEvent::ToolInputDelta`.

File: crates/librefang-llm-drivers/src/drivers/openai.rs [hunk 2]
```diff
@@ -1640,16 +1640,12 @@ impl LlmDriver for OpenAIDriver {
...
-                // Persist the lockout (honors RPH > RPM > retry-after >
-                // 5min default precedence) and reuse the parsed
-                // retry-after for the in-process backoff.
-                let retry_after = crate::shared_rate_guard::record_429_from_headers(
-                    guard_provider,
-                    &guard_key_id,
-                    resp.headers(),
-                    &format!("HTTP 429 from {}", self.base_url),
-                );
                 if attempt < max_retries {
```
Comment (houko) [inline crates/librefang-llm-drivers/src/drivers/openai.rs:1643]: "Retries exhausted" is implemented here as "the terminal attempt was itself a 429", which is narrower — a 429 followed by a *different* terminal failure now persists nothing at all.

Concretely, with `max_retries = 3`:

- attempt 0 -> 429 carrying `x-ratelimit-remaining-requests-1h: 0`, `x-ratelimit-reset-requests-1h: 3540`; `note_429_from_headers`, no lockout
- attempt 1..3 -> connection reset (`transport_error_is_retryable`) or 500, or an un-recovered 400
- returns `LlmError::Http` / `LlmError::Api` from a path that never touches the rate guard

The hour bucket really is exhausted for the next 59 minutes, and before this PR that was on disk. Now nothing is recorded, so this process and every sibling re-runs the full four-request ladder on every subsequent turn until a 429 happens to land on the last attempt.

Same gap at `openai.rs:2104` (stream), `anthropic.rs:549` / `730`, and `gemini.rs:1006` / `1168`.

Suggest holding the last-seen 429 `HeaderMap` (or its parsed `RateLimitSnapshot`) in a local and persisting via `record_from_snapshot` on any terminal `Err` from the loop, rather than only on the 429-shaped return.

File: crates/librefang-llm-drivers/src/drivers/openai.rs [hunk 3]
```diff
@@ -5542,6 +5553,61 @@ mod tests {
         ));
     }
 
+    /// Regression: a tool_call delta for an already-seen lower index must not discard the higher slots accumulated after it.
+    /// The accumulator grows with `resize_with`, which shrinks as well as grows, so the trailing `arguments` fragment for index 0 below used to truncate the Vec back to one element and lose the second parallel call entirely.
+    #[tokio::test]
+    async fn streamed_tool_call_delta_for_earlier_index_keeps_later_calls() {
+        let sse_body = "data: {\"choices\":[{\"delta\":{\"tool_calls\":[{\"index\":0,\"id\":\"call_a\",\"function\":{\"name\":\"first_tool\",\"arguments\":\"{\\\"x\\\":1}\"}}]}}]}\n\
+             data: {\"choices\":[{\"delta\":{\"tool_calls\":[{\"index\":1,\"id\":\"call_b\",\"function\":{\"name\":\"second_tool\",\"arguments\":\"{\\\"y\\\":2}\"}}]}}]}\n\
+             data: {\"choices\":[{\"delta\":{\"tool_calls\":[{\"index\":0,\"function\":{\"arguments\":\"\"}}]}}]}\n\
```
Comment (houko) [inline crates/librefang-llm-drivers/src/drivers/openai.rs:5562]: This frame carries `"arguments":""`, which is a no-op on the accumulator (`push_str("")`), so the test covers only the bare truncation — not the "trailing `arguments` fragment" case the changelog fragment names.

It also means the test cannot observe the emission side: an empty fragment leaves `arguments_emitted == arguments.len()`, so no `ToolInputDelta` is produced and the out-of-order-delta attribution problem stays invisible.

Splitting index 0's arguments across the two frames (e.g. `{"x":` in the first frame and `1}` here) exercises both the accumulator and the event stream, and keeps the `tool_calls[0].input["x"] == 1` assertion below meaningful.
